### PR TITLE
add inverted index classifier

### PR DESCRIPTION
**Pull request by TkrUdagawa**
related to #234 .
I have added classifier 2 methods using inverted index storage.
One uses cosine similarity, the other uses euclidean distance for calculating the score.
Note that current implementation doesn't support deleting label and unlearner.

### CODE REVIEW

File: jubatus/core/classifier/nearest_neighbor_classifier_util.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
+// Jubatus: Online machine learning framework for distributed environment
```
Comment (kazuki) [inline jubatus/core/classifier/nearest_neighbor_classifier_util.cpp]: make_id_from_label/get_labe_from_id is  already defined in nearest_neighbor_classifier.cpp.
please remove same function from nearest_neighbor_classifier.cpp and change to call a nearest_neighbor_classifier_util's functions.

Comment (TkrUdagawa) [inline jubatus/core/classifier/nearest_neighbor_classifier_util.cpp]: Thanks. I have removed the code from nearest_neighbor_classifier.cpp.

**Comment by kazuki** [comment]
20news 4-fold CV (nearest_neighbor_num=19, local_sensitivity=1)
* This pullreq: train=0.604626[ms], classify=1.41695[ms], F=0.841176
* [NN-based InvertedIndex](https://github.com/kazuki/jubatus_core/tree/wip/nn_inverted_index): train=0.575518[ms], classify=1.19541[ms], F=0.863579

This result is correct ?

File: jubatus/core/classifier/cosine_similarity_classifier.cpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    size_t k,
+    float alpha) : inverted_index_classifier(k, alpha) {
+}
+
+void cosine_similarity_classifier::classify_with_scores(
+    const common::sfv_t& fv,
+    classify_result& scores) const {
+  std::vector<std::pair<std::string, float> > ids;
+  mixable_storage_->get_model()->calc_scores(fv, ids, k_);
+
```
Comment (kazuki) [inline jubatus/core/classifier/cosine_similarity_classifier.cpp]: note: missing? https://github.com/jubatus/jubatus_core/blob/master/jubatus/core/recommender/inverted_index.cpp#L82

Comment (TkrUdagawa) [inline jubatus/core/classifier/cosine_similarity_classifier.cpp]: Thanks. You are right.
I need to modify the value as nearer point has lower score.

File: jubatus/core/classifier/cosine_similarity_classifier.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef JUBATUS_CORE_CLASSIFIER_COSINE_SIMILARITY_CLASSIFIER_HPP_
+#define JUBATUS_CORE_CLASSIFIER_COSINE_SIMILARITY_CLASSIFIER_HPP_
+
+#include <string>
+#include "inverted_index_classifier.hpp"
+
+namespace jubatus {
+namespace core {
+namespace classifier {
+class cosine_similarity_classifier : public inverted_index_classifier {
```
Comment (kmaehashi) [inline jubatus/core/classifier/cosine_similarity_classifier.hpp]: (coding style) need blank line here

File: jubatus/core/classifier/euclidean_distance_classifier.hpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+#ifndef JUBATUS_CORE_CLASSIFIER_EUCLIDEAN_DISTANCE_CLASSIFIER_HPP_
+#define JUBATUS_CORE_CLASSIFIER_EUCLIDEAN_DISTANCE_CLASSIFIER_HPP_
+
+#include <string>
+#include "inverted_index_classifier.hpp"
+
+namespace jubatus {
+namespace core {
+namespace classifier {
+class euclidean_distance_classifier : public inverted_index_classifier {
```
Comment (kmaehashi) [inline jubatus/core/classifier/euclidean_distance_classifier.hpp]: (coding style) need blank line here

File: jubatus/core/classifier/inverted_index_classifier.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+void inverted_index_classifier::train(
+    const common::sfv_t& fv,
+    const std::string& label) {
+  std::string id;
+  {
+    util::concurrent::scoped_lock lk(rand_mutex_);
+    id = make_id_from_label(label, rand_);
+  }
+  storage::inverted_index_storage& inv = *mixable_storage_->get_model();
```
Comment (kmaehashi) [inline jubatus/core/classifier/inverted_index_classifier.cpp]: Need to take an exclusive lock here?

**Comment by kmaehashi** [comment]
As `inverted_index_storage` is thread-unsafe, all accesses to the model (i.e., `mixable_storage_->get_model()`) should be protected with read/write lock in classifier layer.

**Comment by TkrUdagawa** [comment]
Thanks. I have added read/write locks.

**Comment by kmaehashi** [comment]
:+1: LGTM

**Comment by kmaehashi** [comment]
Directly fixed include guard error in commit 57fd558.

**Pull request merged**